### ISSUE TITLE
Using top-level programs results in analyzer failures

### ISSUE THREAD

**Issue by bkoelman** [dotnet/roslyn#58120]
**Version Used**: 
Microsoft.CodeAnalysis v4.0.1

**Steps to Reproduce**:
Consider the code below, taken from an existing analyzer. This analyzer is designed to report a diagnostic when a developer declares a type without putting it in a namespace:
```c#
private static void AnalyzeNamedType(SymbolAnalysisContext context)
{
    var type = (INamedTypeSymbol)context.Symbol;

    if (type.ContainingNamespace.IsGlobalNamespace && type.Locations.Any())
    {
        context.ReportDiagnostic(Diagnostic.Create(GlobalTypeRule, type.Locations[0], type.Name, type.ContainingAssembly.Name));
    }
}
```

The analyzer incorrectly reports a diagnostic in a top-level program: "Type 'Program' is declared in the global namespace, which does not match with assembly name 'MyCompany.MyProduct'.". This analyzer should not report a diagnostic for a top-level program, because the user has not declared the Program type. But it should keep reporting a diagnostic when the user has written code for a class named "Program" that is in the global namespace.

As the author of the analyzer, I'd like to fix this bug. But there does not seem to be an API to distinguish between there, as `WellKnownMemberNames.TopLevelStatementsEntryPointTypeName` equals "Program", a location is reported for a symbol that does not exist in the source code and `type.IsImplicitlyDeclared` returns `false`. This doesn't make any sense to me.

**Expected Behavior**:
`type.Locations.Any()` to return empty and `type.IsImplicitlyDeclared` to return `true`, or another API to determine whether the Program class is compiler-generated or explicitly defined in source code.

**Actual Behavior**:
`type.Locations.Any()` returns the location of the first statement within the top-level file, which is something else entirely.

**Comment by Youssef1313** [comment]
You can use `type.GetMembers(WellKnownMemberNames.TopLevelStatementsEntryPointMethodName)` to determine whether the given symbol is a top-level class.
The entry-point method name is still unspeakable, ie, you can't declare one with the same name in source.

**Comment by bkoelman** [comment]
Thanks, that works. I looked at `type.MemberNames`, which returns an empty collection.

**Comment by CyrusNajmabadi** [comment]
Sounds like this has been answered.  Thanks @Youssef1313  :)
